### PR TITLE
Use host ntp server

### PR DESCRIPTION
**Pull request by ogaca-dd**
### What does this PR do?

NTP check now has an option to use local defined NTP server in the host.

### Motivation

https://github.com/DataDog/integrations-core/issues/1816

### Additional Notes

Please check if NTP server can be configured in a different way on our supported OSes.

### CODE REVIEW

File: cmd/agent/dist/conf.d/ntp.d/conf.yaml.default [hunk 1]
```diff
@@ -17,3 +17,8 @@ instances:
     # port: 123
     # version: 3
     # timeout: 5
+    #    
+    # Use the ntp servers defined in the host.    
+    # For Unix system, the servers defined in /etc/ntp.conf and etc/xntp.conf are used.
+    # For Windows system, the servers defined in registry key HKLM\SYSTEM\CurrentControlSet\Services\W32Time\Parameters\NtpServer are used.
```
Comment (olivielpeau) [inline cmd/agent/dist/conf.d/ntp.d/conf.yaml.default]: this should be reflected in integrations-core

Comment (olivielpeau) [reply cmd/agent/dist/conf.d/ntp.d/conf.yaml.default]: let's document as a caveat that enabling this option will make the ntp check unable to detect issues if the clock of the target NTP server of the system is skewed

File: pkg/collector/corechecks/net/ntp_local_defined_server_nowindows.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+				line = strings.TrimSpace(line)
+				fields := strings.Fields(line)
+				if len(fields) >= 2 && fields[0] == "server" {
+					serversMap[fields[1]] = true
+				}
+			}
+		}
+	}
+
+	if len(serversMap) == 0 {
```
Comment (arbll) [inline pkg/collector/corechecks/net/ntp_local_defined_server_nowindows.go]: You are doing this check here but not in the windows implementation. Let's align everything

File: pkg/collector/corechecks/net/ntp.go [hunk 3]
```diff
@@ -78,7 +79,19 @@ func (c *ntpConfig) parse(data []byte, initData []byte) error {
...
+	if c.instance.UseLocalDefinedServers {
+		localNtpServers, err = getLocalServers()
+		if err != nil {
+			return err
+		}
+		log.Infof("Use local defined servers: %v", localNtpServers)
+	}
+
+	if len(localNtpServers) > 0 {
+		c.instance.Hosts = localNtpServers
```
Comment (arbll) [inline pkg/collector/corechecks/net/ntp.go]: The behavior is inconsistent here between windows and linux. If you find 0 ntp server on linux `parse` will error out (see my other comment). On windows you will get default servers. Is this intended ?

**Review by arbll — CHANGES_REQUESTED** [review]
Looks good, just a small consistency issue

File: pkg/collector/corechecks/net/ntp_local_defined_server_nowindows.go (hunk 2 above)
Comment (ogaca-dd) [reply pkg/collector/corechecks/net/ntp_local_defined_server_nowindows.go]: Fixed.

File: pkg/collector/corechecks/net/ntp.go (hunk 3 above)
Comment (ogaca-dd) [reply pkg/collector/corechecks/net/ntp.go]: Fixed.

File: cmd/agent/dist/conf.d/ntp.d/conf.yaml.default (hunk 1 above)
Comment (ogaca-dd) [reply cmd/agent/dist/conf.d/ntp.d/conf.yaml.default]: https://github.com/DataDog/integrations-core/pull/4388

**Review by arbll — APPROVED** [review]
LGTM

**Comment by debraj-manna** [comment]
Is there any plan for making the same fix for agent v5?

**Comment by olivielpeau** [comment]
Hi @debraj-manna, we are not planning to port this enhancement to the Agent v5 at the moment. 

Please consider upgrading to Agent v6 (see our [docs](https://docs.datadoghq.com/agent/guide/upgrade-to-agent-v6/)), if there's anything preventing you from upgrading, or if you need any further guidance feel free to reach out to our support team at support@datadoghq.com.

Thanks!

**Comment by daenney** [comment]
This implementation is problematic for anyone running an OS that uses systemd-timesyncd (which uses a unit style file that may exist at `/etc/systemd/timesyncd` otherwise uses it's compile time defaults), or Chrony, which defaults to configuration in `/etc/chrony.conf`.

The solution here won't work for Google Container-Optimized OS machines, or CoreOS for example, which are fairly common OS targets for cloud.

**Comment by daenney** [comment]
Looking a bit further, it seems that's already possible with https://github.com/DataDog/datadog-agent/blob/master/cmd/agent/dist/conf.d/ntp.d/conf.yaml.default perhaps, by overriding the `host` value? Is that `X` in there required?

**Pull request merged**